parsers/ELEXON.py

```diff
@@ -0,0 +1,239 @@
+#!/usr/bin/env python3
+# coding=utf-8
+
+"""
+Parser that uses the ELEXON API to return the following data types.
+
+Production
+Exchanges
+
+Documentation:
+https://www.elexon.co.uk/wp-content/uploads/2017/06/
+bmrs_api_data_push_user_guide_v1.1.pdf
+"""
+
+import os
+import arrow
+import logging
+import requests
+import datetime as dt
+import pandas as pd
+from io import StringIO
+
+ELEXON_ENDPOINT = 'https://api.bmreports.com/BMRS/{}/v1'
+
+REPORT_META = {
+    'B1620': {
+        'expected_fields': 13,
+        'skiprows': 5
+    },
+    'INTERFUELHH': {
+        'expected_fields': 7,
+        'skiprows': 0
+    }
+}
+
+# 'hydro' key is for hydro production
+# 'hydro storage' key is for hydro storage
+RESOURCE_TYPE_TO_FUEL = {
+    'Biomass': 'biomass',
+    'Fossil Gas': 'gas',
+    'Fossil Hard coal': 'coal',
+    'Fossil Oil': 'oil',
+    'Hydro Pumped Storage': 'hydro storage',
+    'Hydro Run-of-river and poundage': 'hydro',
+    'Nuclear': 'nuclear',
+    'Solar': 'solar',
+    'Wind Onshore': 'wind',
+    'Wind Offshore': 'wind',
+    'Other': 'unknown'
+}
+
+EXCHANGES = {
+    'FR->GB': 3,
+    'GB-NIR->IE': 4,
+    'GB->NL': 5,
+    'GB->IE': 6
+}
+
+
+def query_ELEXON(report, session, params):
+    if 'ELEXON_TOKEN' not in os.environ:
+        raise Exception('No ELEXON_TOKEN found! Please add it to secrets.env!')
+    params['APIKey'] = os.environ['ELEXON_TOKEN']
+    return session.get(ELEXON_ENDPOINT.format(report), params=params)
+
+
+def query_exchange(session, target_datetime=None):
+    if target_datetime is None:
+        target_datetime = dt.date.today()
+
+    from_date = (target_datetime - dt.timedelta(days=1)).strftime('%Y-%m-%d')
+    to_date = target_datetime.strftime('%Y-%m-%d')
+
+    params = {
+        'FromDate': from_date,
+        'ToDate': to_date,
+        'ServiceType': 'csv'
+    }
+    response = query_ELEXON('INTERFUELHH', session, params)
+    return response.text
+
+
+def query_production(session, target_datetime=None):
+    if target_datetime is None:
+        target_datetime = dt.date.today()
+
+    params = {
+        'SettlementDate': target_datetime.strftime('%Y-%m-%d'),
+        'Period': '*',
+        'ServiceType': 'csv'
+    }
+    response = query_ELEXON('B1620', session, params)
```

Comment:
@maxbellec it seems here that ELEXON sends data back for D to D+1.
I think this is in contradiction with others APIs, where data is fetch *up to* `target_datetime`.
Will this be a problem?

Reply:
indeed, if the target datetime corresponds to midnight, the parser should fetch the day before. Else, it should fetch the requested day.